### PR TITLE
Ignore unmanaged schemas

### PR DESCRIPTION
**Pull request by aheyman11**
Defines the concept of "managed" schemas to be those specified in your DW config, and allow Arthur to ignore unmanaged schemas 1) when determining dependency ordering, and 2) when loading a relation with unmanaged dependencies in staging mode.

The most prominent change is that RelationDescription dependencies are now TableName objects instead of strings, and those objects have an `is_managed` property that references `etl.config`.

### CODE REVIEW

File: python/etl/names.py [hunk 1]
```diff
@@ -84,25 +87,28 @@ class TableName:
...
     >>> staging_purchases = purchases.as_staging_table_name()
     >>> staging_purchases.table == purchases.table
     True
     >>> staging_purchases.schema == purchases.schema
     False
     """
 
-    __slots__ = ("_schema", "_table", "_staging")
+    __slots__ = ("_schema", "_table", "_staging", "_base_schemas")
+    _identifier_slots = ("_schema", "_table", "_staging")
```
Comment (bhtucker) [inline python/etl/names.py]: It seems like we use `.identifier` as a sufficient hash key — can we drop `_staging` out of here and then use these slots in `to_tuple`?

File: python/etl/names.py [hunk 2]
```diff
@@ -84,25 +87,28 @@ class TableName:
...
 
     def __init__(self, schema: Optional[str], table: str) -> None:
         # Concession to subclasses ... schema is optional
         self._schema = schema.lower() if schema else None
         self._table = table.lower()
         self._staging = False
+        self._base_schemas = None  # type: Optional[List]
```
Comment (bhtucker) [inline python/etl/names.py]: If we only care to do string containment checks, we should go for a set of strings

File: python/etl/names.py [hunk 3]
```diff
@@ -137,6 +149,10 @@ def identifier(self) -> str:
         """
         return "{}.{}".format(*self.to_tuple())
 
+    @property
+    def is_managed(self) -> bool:
+        return self._schema in [s.name for s in self.base_schemas]
```
Comment (bhtucker) [inline python/etl/names.py]: In keeping with the set-of-strings idea, let `base_schemas` access the `name` property

File: python/etl/names.py [hunk 4]
```diff
@@ -242,6 +260,7 @@ def match_pattern(self, pattern: str) -> bool:
     def as_staging_table_name(self):
         tn = TableName(*self.to_tuple())
         tn._staging = True
+        tn._base_schemas = self._base_schemas
```
Comment (bhtucker) [inline python/etl/names.py]: We do save a little execution time by copying this, but I'd advocate for the simplicity of only setting the private property mentioned in the method name

File: python/etl/relation.py [hunk 5]
```diff
@@ -188,9 +188,10 @@ def query_stmt(self) -> str:
         return str(self._query_stmt)  # The str(...) shuts up the type checker.
 
     @property
-    def dependencies(self) -> FrozenSet[str]:
+    def dependencies(self) -> FrozenSet[TableName]:
         if self._dependencies is None:
-            self._dependencies = frozenset(self.table_design.get("depends_on", []))
+            dependent_table_names = [TableName.from_identifier(d) for d in self.table_design.get("depends_on", [])]
+            self._dependencies = frozenset(dependent_table_names)
```
Comment (bhtucker) [inline python/etl/relation.py]: This saves us a `TableName` instantiation in the SQL rewrite for staging

File: python/etl/relation.py [hunk 6]
```diff
@@ -401,7 +404,7 @@ def order_by_dependencies(relation_descriptions):
         minimum, tie_breaker, description = queue.get()
         if minimum > 2 * nr_tables:
             raise CyclicDependencyError("Cannot determine order, suspect cycle in DAG of dependencies")
-        others = [table_map[dep].order for dep in description.dependencies]
+        others = [table_map[dep.identifier].order for dep in description.dependencies if dep.is_managed]
```
Comment (bhtucker) [inline python/etl/relation.py]: We can keep the pattern of sets/dicts of `TableName`s if we make `table_map` out of `target_table_name`s

File: python/etl/load.py [hunk 7]
```diff
@@ -112,7 +112,10 @@ def __format__(self, code):
         the position (staging or not) of a table.
 
         >>> import etl.file_sets
-        >>> fs = etl.file_sets.TableFileSet(TableName("a", "b"), TableName("c", "b"), None)
+        >>> from etl.config.dw import DataWarehouseSchema
+        >>> target_table_name = TableName("c", "b")
+        >>> target_table_name._base_schemas = [DataWarehouseSchema(schema_info={'name': 'c', 'owner': ''})]
```
Comment (tvogels01) [inline python/etl/load.py]: Once we start using `_base_schemas` outside the `names` module, accessing something private is not so great ...
At a minimum, we could add a setter:
```
    @base_schemas.setter
    def base_schemas(self, schemas: List) -> None:
        self._base_schemas = schemas
```
Theoretically, we could make sure that the base_schemas get set only once (which is the assumption in code, I believe).  But practically, this feels outside the scope.

File: python/etl/names.py (hunk 1 above)
Comment (tvogels01) [reply python/etl/names.py]: On a separate yet related note ... I'm not a fan of `_identifier_slots`.  The `__slots__` tuple is more of a setting than a field as it is a hint to Python which field names will be used. But the "identifier slots" are really more of a class constant (and are used only once). So it's more `HASH_FIELDS` or `UNIQUE_IDENTIFIER_FIELDS`?

Could we just use `to_tuple`? (`hash(self.to_tuple())`?) Conceptually, could we ever talk about tables that are both in staging and not at the same time? If the answer is yes, then simply using `__slots__` in the `hash` method is semantically meaningful. (This last point is how I read Benson's comment.)

File: python/etl/names.py (hunk 2 above)
Comment (tvogels01) [reply python/etl/names.py]: Not quite.  Please keep this as a list. The "base schemas" also allow sorting and make the list of all tables predictable (which is nicely hidden in the selector code.) Using a set would kill that since changing the config would alter the order of schemas when printed. If we're worried about performance, we could use an `OrderedDict`.

Overall (my) expectation is that the list of relations is ordered by the schemas in the way they are ordered in the configuration and then alphabetically by table name. Since sources come before transformations, this keeps us from mingling source and transformation schemas in ordering.

File: python/etl/names.py [hunk 8]
```diff
@@ -84,25 +87,28 @@ class TableName:
...
         self._schema = schema.lower() if schema else None
         self._table = table.lower()
         self._staging = False
+        self._base_schemas = None  # type: Optional[List]
 
     @property
     def schema(self):
-        # for system table dependencies, the schema should not be in a "staging" version
-        if self.staging and not self._schema.startswith('pg_catalog'):
+        # for unmanaged dependencies, the schema should not be in a "staging" version
```
Comment (tvogels01) [inline python/etl/names.py]: This comment should probably start going into the documentation of the class. This interface of "staging" versions is official enough now (and getting more complex) to warrant an explanation in the class doc.

File: python/etl/names.py [hunk 9]
```diff
@@ -115,6 +121,12 @@ def table(self):
     def staging(self):
         return self._staging
 
+    @property
+    def base_schemas(self) -> List:
+        if self._base_schemas is None:
+            self._base_schemas = etl.config.get_dw_config().schemas
+        return self._base_schemas or []  # the 'or []' is just for the type checker
```
Comment (tvogels01) [inline python/etl/names.py]: Torn between "nifty" and "horrified" :-)   Actually, this expresses our idea that "empty means anything" so wouldn't be there just for the type checker. Going with "nifty".

File: python/etl/names.py (hunk 3 above)
Comment (tvogels01) [reply python/etl/names.py]: Curious ... if "base schemas" are not set then everything is "not managed"? This changes past assumptions. Ok, but worth pointing out.

(We'll need something like "any schemas not defined as source or transformation in the configuration will be assumed to be outside of Arthur's control and thus not managed.")

File: python/etl/names.py (hunk 4 above)
Comment (tvogels01) [reply python/etl/names.py]: This deservers a comment (doc string) to understand that his is a cloning method.

Purists might argue that we should have something like a `staging: Optional[bool]=False` argument in the constructor to make it easier to see what's going on. Then this method would simply be (along with Benson's observation):
```
def as_staging_table_name(self):
   return TableName(*self.to_tuple(), staging=True)
```

File: python/etl/relation.py [hunk 10]
```diff
@@ -364,7 +365,7 @@ def order_by_dependencies(relation_descriptions):
     RelationDescription.load_in_parallel(relation_descriptions)
     descriptions = [SortableRelationDescription(description) for description in relation_descriptions]
 
-    known_tables = frozenset({description.identifier for description in descriptions})
+    known_tables = frozenset({d.target_table_name for d in relation_descriptions})
```
Comment (tvogels01) [inline python/etl/relation.py]: Is there a reason to go from `description` to `d`? I find code that has the meaning spelled out for me so much more readable. And bits are mostly free these days ;-)
Either
```
    known_tables = frozenset({description.target_table_name for description in relation_descriptions})
```
or
```
    known_tables = frozenset({relation.target_table_name for relation in relation_descriptions})
```
is harder to write but much easier to read in the long run. This is true here and in a couple other places in this PR (for some value of "true" in my own opinion).

File: python/etl/relation.py [hunk 11]
```diff
@@ -373,25 +374,27 @@ def order_by_dependencies(relation_descriptions):
     known_unknowns = set()
     queue = PriorityQueue()
     for initial_order, description in enumerate(descriptions):
-        pg_internal_dependencies = set(d for d in description.dependencies if d.startswith('pg_catalog'))
-        unknowns = description.dependencies - known_tables - pg_internal_dependencies
+        # superset including internal dependencies
+        unmanaged_dependencies = set(d for d in description.dependencies if not d.is_managed)
+        pg_internal_dependencies = set(d for d in description.dependencies if d.schema == 'pg_catalog')
```
Comment (tvogels01) [inline python/etl/relation.py]: If the lack of symmetry offends you, an `is_in_catalog` property would help. (I'm fine with as-is.)

File: python/etl/relation.py (hunk 6 above)
Comment (tvogels01) [reply python/etl/relation.py]: Along that line ... should `target_table_name` be a field in a sortable relation description? Reaching into the "original description" feels a bit wrong -- the `original_description` is there but ideally the algorithm gets away with just using fields immediately on the sortable relation description to keep the up the appearance of the facade.

File: python/etl/names.py (hunk 2 above)
Comment (tvogels01) [reply python/etl/names.py]: Appending my original comment ... The problem that I see is that we need to avoid having the same name for different things (and my comment is an example of that confusion). So for me, the base schemas are the list of schema names in order of the config. If we want to have something to easily check against, then a set is great but should be named for what it is used for. Keeping `_base_schemas` and changing the type will get confusing. Since this field is used to detect "managed" schemas, should it then be called just that?
```
        ...
        self._managed_schemas = frozenset()

    @managed_schemas.setter
    def managed_schemas(self, schemas: List) -> None:
        self._managed_schemas = frozenset(schemas)
```

File: python/etl/names.py (hunk 1 above)
Comment (aheyman11) [reply python/etl/names.py]: I can't find any instances where we would be comparing staging and non-staging versions of the same table, so I'm going to redefine the hash function as `hash(self.to_tuple())`.

FWIW, the original motivation behind excluding `_base_schemas` from the hash was only that that object was a list and thus non-hashable.

File: python/etl/names.py (hunk 2 above)
Comment (aheyman11) [reply python/etl/names.py]: I hadn't realized that we also use base schemas in `TableSelector`s. I've renamed the new concept to `managed_schemas`, which is now a frozenset of strings.

File: python/etl/relation.py [hunk 12]
```diff
@@ -373,25 +374,27 @@ def order_by_dependencies(relation_descriptions):
...
+        unknowns = description.dependencies - known_tables - unmanaged_dependencies
         if unknowns:
             known_unknowns.update(unknowns)
-            has_unknown_dependencies.add(description.identifier)
-            # Drop the unknowns from the list of dependencies so that the loop below doesn't wait for their resolution.
-            description.dependencies = description.dependencies.difference(unknowns)
+            has_unknown_dependencies.add(description.original_description.target_table_name)
+        if unmanaged_dependencies:
+            logger.info("The following dependent relations are not managed by Arthur: %s",
+                        join_with_quotes(unmanaged_dependencies))
```
Comment (tvogels01) [inline python/etl/relation.py]: The `unmanaged_dependencies` here are `List[TableName]`? If so then, it would be nice to unpack their identifiers like in the other occasions when using `join_with_quotes`.
Also, are these "dependent relations"? Aren't these relations that this relation depends on?

File: python/etl/relation.py (hunk 5 above)
Comment (aheyman11) [reply python/etl/relation.py]: Fixed

File: python/etl/relation.py [hunk 13]
```diff
@@ -373,25 +374,27 @@ def order_by_dependencies(relation_descriptions):
...
-            has_internal_dependencies.add(description.identifier)
+            has_internal_dependencies.add(description.original_description.target_table_name)
         queue.put((1, initial_order, description))
     if has_unknown_dependencies:
         logger.warning("These relations were unknown during dependency ordering: %s",
-                       join_with_quotes(known_unknowns))
+                       join_with_quotes([d.identifier for d in known_unknowns]))
         logger.warning('This caused these relations to have dependencies that are not known: %s',
-                       join_with_quotes(has_unknown_dependencies))
+                       join_with_quotes([d.identifier for d in has_unknown_dependencies]))
```
Comment (tvogels01) [inline python/etl/relation.py]: Should we upgrade `join_with_quotes` to automatically use the identifier?
```python
    if isinstance(names, (set, frozenset)):
        if all(hasattr(name, "identifier") for name in names):
            return ', '.join(sorted("'{}'".format(name.identifier) for name in names))
        else:
            return ', '.join("'{}'".format(name) for name in sorted(names))
    else:
        return ', '.join("'{}'".format(name) for name in names)
```

**Comment by aheyman11** [comment]
Besides the `join_with_quotes` discussion, I *believe* that all other comments have been addressed.

File: python/etl/names.py [hunk 14]
```diff
@@ -115,6 +127,20 @@ def table(self):
...
 
+    @property
+    def managed_schemas(self) -> frozenset:
+        if self._managed_schemas is not None:
+            return self._managed_schemas
+        try:
+            schemas = etl.config.get_dw_config().schemas
+        except AttributeError:
+            raise ETLSystemError("dw_config has not been set!")
+        return frozenset(schema.name for schema in schemas)
```
Comment (bhtucker) [inline python/etl/names.py]: Does this ever set `self._managed_schemas` to be the result here? Not sure if it's left out or there's some magic I'm missing
